ARM: Cortex-M: add Cortex-M23 board target

ioannisg opened the pull request:
Completes build coverage for Cortex-M variants :)

Follow up: add TF-M integration for Baseline Cortex-M, since this is already a supported emulator in TF-M.
Possibly requires assembly adaptations in cortex-m code.

ioannisg committed:
soc: arm: add definition for AN519 SoC

SoC definition for MPS2 AN519.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: board definition for mps2_an519

Adding board definition for mps2 an519, which
is a Cortex-M23 implementation.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

galak commented on boards/arm/mps2_an519/Kconfig.board:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_MPS2_AN519
+	bool "Arm Cortex-M23 (Corstone-300, SSE-300 w/Ethos-U55) on MPS3 (AN547)"
```
need to fixup the comment string here

ioannisg replied:
done

gmarull commented on boards/arm/mps2_an519/mps2_an519_defconfig:
```diff
@@ -0,0 +1,31 @@
+# Build a Secure firmware image
+CONFIG_TRUSTED_EXECUTION_SECURE=y
+
+CONFIG_I2C=y
```
is i2c needed by default config?

ioannisg replied:
nope. removed

gmarull commented on boards/arm/mps2_an519/mps2_an519-common.dtsi:
```diff
@@ -0,0 +1,205 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+ sysclk: system-clock {
```
```suggestion
sysclk: system-clock {
```

ioannisg replied:
done, thanks.

gmarull commented on boards/arm/mps2_an519/doc/index.rst:
```diff
@@ -0,0 +1,231 @@
+a MPS3 AN547 hardware platform.
+
+The Corstone-300 FVP (Fixed Virtual Platform) is a complete simulation of the
+Arm system, including processor, memory and peripherals. It is a available free
```
```suggestion
Arm system, including processor, memory and peripherals. It is available free
```

ioannisg replied:
This part was removed, so this is resolved.

ioannisg committed:
arm: cortex-m: fix assembly code in swap_helper.S for Cortex-M23

Fix some assembly code-blocks in swap_helper.S so the code
can compile for Cortex-M23. No behavioral changes for other
Cortex-M variants.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arm: cortex-m: fix assembly code in userspace.S for Cortex-M23

Fix some assembly code-blocks in userspace.S so the code
can compile for Cortex-M23. No behavioral changes for other
Cortex-M variants. This mostly concerns code blocks that
implement stack limit checking for Cortex-M23.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
doc: release: add entry for mps2_an519 in release notes for 2.7

Add an entry in the v2.7 release notes for adding mps2_an519.
Add a note about supporting stack limit checking for Cortex-M23.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

gmarull reviewed (approved):
A couple of minor nitpicks, +1 for board and docs part

ioannisg commented:
@gmarull will address your comments, thanks.
@galak @microbuilder could you take a brief look? 
@galak comments on the assembly fixes are welcome.

ioannisg committed:
boards: mps2: minor doc fixes

Set code-block type from bash to none in
mps2_an521 and mps3_an547 board docs, in
places where the corresponding note does
not concern terminal input.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented on boards/arm/mps2_an519/doc/index.rst:
```diff
@@ -0,0 +1,374 @@
+2. Open ``<MPS2 device name>/MB/HBI0263C/AN519/images.txt``.
+3. Update the ``AN519/images.txt`` file as follows:
+
+.. code-block:: bash
```
Fixed.
Will add a separate commit to update this (as this was a copy past from another mps2 .rst)

ioannisg commented on boards/arm/mps2_an519/doc/index.rst:
```diff
@@ -0,0 +1,374 @@
+For mode details refer to `MPS2+ AN519 Technical Reference Manual (TRM)`_.
+
+LED
+============
```
fixed, thanks.

microbuilder commented:
I've done a quick build test, but it's unfortunate that QEMU doesn't support this board yet for a quick test run, and I didn't see any M23 boards in the list:

```bash
$ qemu-system-arm -machine help | grep mps2
mps2-an385           ARM MPS2 with AN385 FPGA image for Cortex-M3
mps2-an386           ARM MPS2 with AN386 FPGA image for Cortex-M4
mps2-an500           ARM MPS2 with AN500 FPGA image for Cortex-M7
mps2-an505           ARM MPS2 with AN505 FPGA image for Cortex-M33
mps2-an511           ARM MPS2 with AN511 DesignStart FPGA image for Cortex-M3
mps2-an521           ARM MPS2 with AN521 FPGA image for dual Cortex-M33
```

microbuilder reviewed (approved):
Confirmed with a build test. Limited means to test further with no QEMU support available, at present.

cfriedt commented:
@ioannisg - the test that is currently failing in CI passes when rebased on main. Can you please rebase?

```
zephyr % twister -p mps2_an521 -s tests/arch/arm/arm_interrupt/arch.interrupt.no_optimizations  
Renaming output directory to /Users/cfriedt/workspace/zephyrproject/zephyr/twister-out.8
INFO    - Zephyr version: v2.7.0-rc3-217-gad916a8c906a
INFO    - JOBS: 8
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testcase list...
INFO    - 1 test scenarios (1 configurations) selected, 0 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:    1/   1  100%  skipped:    0, failed:    0
INFO    - 1 of 1 test configurations passed (100.00%), 0 failed, 0 skipped with 0 warnings in 51.87 seconds
INFO    - In total 4 test cases were executed, 0 skipped on 1 out of total 398 platforms (0.25%)
INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing xunit report /Users/cfriedt/workspace/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /Users/cfriedt/workspace/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

cfriedt committed:
Merge branch 'main' into qemu_cortex_m23

cfriedt commented:
> @ioannisg - the test that is currently failing in CI passes when rebased on main. Can you please rebase?

I was able to do it - should pass CI now 🤞
